From 72d0a6a491a9bc828f540ac8c79f6f91dcacfeec Mon Sep 17 00:00:00 2001
From: Undefined01 <amoscr@163.com>
Date: Fri, 19 Sep 2025 21:31:09 +0800
Subject: [PATCH] Patched for NEMU

---
 p/link.ld      |  1 +
 p/riscv_test.h | 63 ++------------------------------------------------
 2 files changed, 3 insertions(+), 61 deletions(-)

diff --git a/p/link.ld b/p/link.ld
index b3e315e..ab93438 100644
--- a/p/link.ld
+++ b/p/link.ld
@@ -13,5 +13,6 @@ SECTIONS
   .data : { *(.data) }
   .bss : { *(.bss) }
   _end = .;
+  .debug : { *(.note.gnu.build-id) }
 }
 
diff --git a/p/riscv_test.h b/p/riscv_test.h
index 3d4637a..9ace47a 100644
--- a/p/riscv_test.h
+++ b/p/riscv_test.h
@@ -182,70 +182,13 @@
 
 #define RVTEST_CODE_BEGIN                                               \
         .section .text.init;                                            \
-        .align  6;                                                      \
-        .weak stvec_handler;                                            \
-        .weak mtvec_handler;                                            \
         .globl _start;                                                  \
 _start:                                                                 \
-        /* reset vector */                                              \
-        j reset_vector;                                                 \
-        .align 2;                                                       \
-trap_vector:                                                            \
-        /* test whether the test came from pass/fail */                 \
-        csrr t5, mcause;                                                \
-        li t6, CAUSE_USER_ECALL;                                        \
-        beq t5, t6, write_tohost;                                       \
-        li t6, CAUSE_SUPERVISOR_ECALL;                                  \
-        beq t5, t6, write_tohost;                                       \
-        li t6, CAUSE_MACHINE_ECALL;                                     \
-        beq t5, t6, write_tohost;                                       \
-        /* if an mtvec_handler is defined, jump to it */                \
-        la t5, mtvec_handler;                                           \
-        beqz t5, 1f;                                                    \
-        jr t5;                                                          \
-        /* was it an interrupt or an exception? */                      \
-  1:    csrr t5, mcause;                                                \
-        bgez t5, handle_exception;                                      \
-        INTERRUPT_HANDLER;                                              \
-handle_exception:                                                       \
-        /* we don't know how to handle whatever the exception was */    \
-  other_exception:                                                      \
-        /* some unhandlable exception occurred */                       \
-  1:    ori TESTNUM, TESTNUM, 1337;                                     \
-  write_tohost:                                                         \
-        sw TESTNUM, tohost, t5;                                         \
-        sw zero, tohost + 4, t5;                                        \
-        j write_tohost;                                                 \
-reset_vector:                                                           \
         INIT_XREG;                                                      \
-        RISCV_MULTICORE_DISABLE;                                        \
-        INIT_RNMI;                                                      \
-        INIT_SATP;                                                      \
-        INIT_PMP;                                                       \
-        DELEGATE_NO_TRAPS;                                              \
         li TESTNUM, 0;                                                  \
-        la t0, trap_vector;                                             \
-        csrw mtvec, t0;                                                 \
-        CHECK_XLEN;                                                     \
-        /* if an stvec_handler is defined, delegate exceptions to it */ \
-        la t0, stvec_handler;                                           \
-        beqz t0, 1f;                                                    \
-        csrw stvec, t0;                                                 \
-        li t0, (1 << CAUSE_LOAD_PAGE_FAULT) |                           \
-               (1 << CAUSE_STORE_PAGE_FAULT) |                          \
-               (1 << CAUSE_FETCH_PAGE_FAULT) |                          \
-               (1 << CAUSE_MISALIGNED_FETCH) |                          \
-               (1 << CAUSE_USER_ECALL) |                                \
-               (1 << CAUSE_BREAKPOINT);                                 \
-        csrw medeleg, t0;                                               \
-1:      csrwi mstatus, 0;                                               \
         init;                                                           \
         EXTRA_INIT;                                                     \
         EXTRA_INIT_TIMER;                                               \
-        la t0, 1f;                                                      \
-        csrw mepc, t0;                                                  \
-        csrr a0, mhartid;                                               \
-        mret;                                                           \
 1:
 
 //-----------------------------------------------------------------------
@@ -260,21 +203,19 @@ reset_vector:                                                           \
 //-----------------------------------------------------------------------
 
 #define RVTEST_PASS                                                     \
-        fence;                                                          \
         li TESTNUM, 1;                                                  \
         li a7, 93;                                                      \
         li a0, 0;                                                       \
-        ecall
+        ebreak
 
 #define TESTNUM gp
 #define RVTEST_FAIL                                                     \
-        fence;                                                          \
 1:      beqz TESTNUM, 1b;                                               \
         sll TESTNUM, TESTNUM, 1;                                        \
         or TESTNUM, TESTNUM, 1;                                         \
         li a7, 93;                                                      \
         addi a0, TESTNUM, 0;                                            \
-        ecall
+        ebreak
 
 //-----------------------------------------------------------------------
 // Data Section Macro
-- 
2.30.2

