From 35eb7604aa49ddfbe6efe237943bd6a7a3cab33c Mon Sep 17 00:00:00 2001
From: Undefined01 <amoscr@163.com>
Date: Fri, 19 Sep 2025 21:35:03 +0800
Subject: [PATCH] Patched for NEMU

---
 env          | 2 +-
 isa/Makefile | 7 ++++++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/env b/env
index 6de71ed..72d0a6a 160000
--- a/env
+++ b/env
@@ -1 +1 @@
-Subproject commit 6de71edb142be36319e380ce782c3d1830c65d68
+Subproject commit 72d0a6a491a9bc828f540ac8c79f6f91dcacfeec
diff --git a/isa/Makefile b/isa/Makefile
index b1f0c2a..cd42c5e 100644
--- a/isa/Makefile
+++ b/isa/Makefile
@@ -49,6 +49,7 @@ RISCV_PREFIX ?= riscv$(XLEN)-unknown-elf-
 RISCV_GCC ?= $(RISCV_PREFIX)gcc
 RISCV_GCC_OPTS ?= -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles
 RISCV_OBJDUMP ?= $(RISCV_PREFIX)objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data
+RISCV_OBJCOPY ?= $(RISCV_PREFIX)objcopy
 RISCV_SIM ?= spike
 
 vpath %.S $(src_dir)
@@ -65,6 +66,9 @@ vpath %.S $(src_dir)
 %.out32: %
 	$(RISCV_SIM) --isa=rv32gc_ziccid_zfh_zicboz_svnapot_zicntr_zba_zbb_zbc_zbs --misaligned $< 2> $@
 
+%.bin: %
+	$(RISCV_OBJCOPY) --strip-all $^ -O binary $@
+
 define compile_template
 
 $$($(1)_p_tests): $(1)-p-%: $(1)/%.S
@@ -76,8 +80,9 @@ $$($(1)_v_tests): $(1)-v-%: $(1)/%.S
 $(1)_tests += $$($(1)_v_tests)
 
 $(1)_tests_dump = $$(addsuffix .dump, $$($(1)_tests))
+$(1)_tests_bin = $$(addsuffix .bin, $$($(1)_tests))
 
-$(1): $$($(1)_tests_dump)
+$(1): $$($(1)_tests_dump) $$($(1)_tests_bin)
 
 .PHONY: $(1)
 
-- 
2.30.2

